File: scrapers/brooklyn.py
import os
import re
import tempfile
import pytesseract
import subprocess
from pathlib import Path
import csv
from selenium.webdriver import Remote, ChromeOptions
from selenium.webdriver.chromium.remote_connection import ChromiumRemoteConnection
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from pdf2image import convert_from_path
import requests
from datetime import datetime

# Get the AUTH variable from the environment variable BRIGHTDATA_AUTH
AUTH = os.getenv('BRIGHTDATA_AUTH')
PROXY = os.getenv('BRIGHTDATA_PROXY')
if not AUTH:
    raise ValueError("BRIGHTDATA_AUTH environment variable is not set.")
if not PROXY:
    raise ValueError("BRIGHTDATA_PROXY environment variable is not set.")

proxies = {
    'http': PROXY,
    'https': PROXY
}
SBR_WEBDRIVER = f'https://{AUTH}@brd.superproxy.io:9515'

def extract_text_from_pdf(pdf_path):
    # Convert the PDF to a PNG image using pdf2image
    images = convert_from_path(pdf_path, fmt='png')
    image_path = str(pdf_path).replace(".pdf", ".png")
    
    # Save the first page as PNG (assuming single-page PDF)
    images[0].save(image_path, 'PNG')

    # Run Tesseract on the image to extract text
    text = pytesseract.image_to_string(image_path)
    print("-------EXTRACTED PDF --------")
    print(f"{text}")

    return text

def extract_block_lot(text):
    primary_pattern = r"(?i)Block\s*[: ]\s*(\d+)\s*(?:[^\d]*?)Lots?\s*[: ]\s*(\d+)"
    secondary_pattern = r"(\d{3,4})-(\d{1,2})"
    
    match_primary = re.search(primary_pattern, text)
    if match_primary:
        return match_primary.group(1), match_primary.group(2)
    
    match_secondary = re.search(secondary_pattern, text)
    if match_secondary:
        return match_secondary.group(1), match_secondary.group(2)
    
    return None, None

def convert_to_address(filename):
    # Remove the file extension
    base_name = re.sub(r"\.pdf$", "", filename, flags=re.IGNORECASE)
    
    # Replace hyphens with spaces and capitalize each word
    address = " ".join(word.capitalize() for word in base_name.split('-'))
    
    return address

def main():
    # Load existing foreclosure auctions CSV
    existing_auctions_file = 'transactions/foreclosure_auctions.csv'
    existing_auctions = []
    with open(existing_auctions_file, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row['borough'] == 'Brooklyn':
                existing_auctions.append((row['date'], row['case_number']))
    
    print('Connecting to Scraping Browser...')
    sbr_connection = ChromiumRemoteConnection(SBR_WEBDRIVER, 'goog', 'chrome')
    options = ChromeOptions()
    
    # Set download preferences
    download_dir = tempfile.mkdtemp()
    
    with Remote(sbr_connection, options=options) as driver:
        print('Connected! Navigating...')
        
        driver.get('https://www.nycourts.gov/legacyPDFs/courts/2jd/kings/civil/foreclosures/foreclosure%20scans/')
        print('Navigated! Scraping page content...')

        # Extract the auction date from the link text
        html = driver.page_source
        auction_date_match = re.search(r'(\d{1,2}/\d{1,2}/\d{4})', html)
        if not auction_date_match:
            print(f'No auction date found in {link_text}')
            return
        
        auction_date_str = auction_date_match.group(0)
        auction_date = datetime.strptime(auction_date_str, '%m/%d/%Y').strftime('%Y-%m-%d')
        
        # Find all the PDF links on the page and collect their href attributes and link text
        links = driver.find_elements(By.TAG_NAME, 'a')
        pdf_links = [(link.text, link.get_attribute('href')) for link in links if link.get_attribute('href') and link.get_attribute('href').endswith('.pdf')]
        
        
        for link_text, pdf_url in pdf_links:
            # Check if the auction date already exists in the CSV
            if (auction_date, link_text) in existing_auctions:
                print(f'Auction {auction_date}, {link_text} already exists. Skipping.')
                continue

            print(f'Processing PDF: {pdf_url} (Link text: {link_text})')
            
            # Download the PDF using requests
            subprocess.run(['node', 'scrapers/download_pdf.js', pdf_url])
            
            pdf_path = Path(link_text)
            if pdf_path.exists():
                print(f'Extracting text from: {pdf_path}')
                extracted_text = extract_text_from_pdf(pdf_path)
                
                block, lot = extract_block_lot(extracted_text)
                if not block or not lot:
                    print(f'Block and lot not found in {pdf_path}.')
                    continue
                
                address = convert_to_address(link_text)
                
                # Append data to CSV
                with open(existing_auctions_file, mode='a', newline='', encoding='utf-8') as csv_file:
                    writer = csv.writer(csv_file)
                    writer.writerow(['Brooklyn', auction_date, link_text, address, block, lot])
            else:
                print(f'Failed to download PDF: {pdf_url}')

if __name__ == '__main__':
    main()

File: scrapers/brooklyn.js
import { createWriteStream, existsSync } from 'fs';
import { parse } from 'csv-parse';
import { stringify } from 'csv-stringify';
import { readFile, unlink, copyFile } from 'fs/promises';
import { JSDOM } from 'jsdom';
import { download_pdf } from './download_pdf.js';
import {extractTextFromPdf, extractBlockLot, extractIndexNumber, extractJudgement } from './utils.js'
import { connect } from 'puppeteer-core';
import readline from 'readline';
import { exec } from 'child_process';


// Check for the --interactive flag in the command-line arguments
const isInteractive = process.argv.includes('--interactive');

const rl = isInteractive ? readline.createInterface({
    input: process.stdin,
    output: process.stdout
}) : null;

async function prompt(question) {
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            resolve(parseInt(answer));
        });
    });
}


const SBR_WS_ENDPOINT = `wss://${process.env.BRIGHTDATA_AUTH}@brd.superproxy.io:9222`;

function getNextThursday(dateString) {
    // Step 1: Parse the date string into a Date object
    const date = new Date(dateString);

    // Step 2: Calculate the difference in days to the next Thursday
    // Day of the week as an integer (0 = Sunday, 1 = Monday, ..., 6 = Saturday)
    const dayOfWeek = date.getUTCDay();
    
    // Calculate how many days until next Thursday (4)
    const daysUntilNextThursday = (4 - dayOfWeek + 7) % 7 || 7;

    // Step 3: Adjust the date
    date.setUTCDate(date.getUTCDate() + daysUntilNextThursday + 7);

    return date;
}

function convertToAddress(filename) {
    const baseName = filename.replace(/\.pdf$/i, '');
    return baseName.split('-').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
}

async function main() {
    const existingAuctionsFile = 'transactions/foreclosure_auctions.csv';
    let existingAuctionsData = await readFile(existingAuctionsFile, 'utf8');
    const stringifier = stringify({ header: false });
    const writeableStream = createWriteStream(existingAuctionsFile, { flags: 'a' });
    stringifier.pipe(writeableStream);

    let existingAuctions = [];
    parse(existingAuctionsData, {
        columns: true,
        skip_empty_lines: true
    }, (err, records) => {
        if (err) throw err;
        existingAuctions = records.filter(row => row.borough === 'Brooklyn').map(row => [row.date, row.case_number]);
    });

    console.log('Connecting to Scraping Browser...');

    const browser = await connect({
        browserWSEndpoint: SBR_WS_ENDPOINT,
    });

    const page = await browser.newPage();

    try {
        console.log('Connected! Navigating...');
        const url = 'https://www.nycourts.gov/legacyPDFs/courts/2jd/kings/civil/foreclosures/foreclosure%20scans/'
        await page.goto(url, { waitUntil: 'networkidle2' });
        console.log('Navigated! Scraping page content...');

        const html = await page.content();
        const { document } = (new JSDOM(html)).window;
        const auctionDateMatch = document.body.textContent.match(/(\d{1,2}\/\d{1,2}\/\d{4})/);

        if (!auctionDateMatch) {
            console.log(`No auction date found.`);
            return;
        }

        const auctionDateStr = auctionDateMatch[0];
        const adjustedDate = getNextThursday(auctionDateStr)
        const auctionDate = adjustedDate.toISOString().split('T')[0];
        console.log(`Auction date found: ${auctionDate}`)

        const pdfLinks = Array.from(document.querySelectorAll('a'))
            .map(link => [link.textContent, link.href])
            .filter(([_, href]) => href && href.endsWith('.pdf'));

        for (const [linkText, pdfUrl] of pdfLinks) {
            if (existingAuctions.some(([date, caseNumber]) => date === auctionDate && caseNumber === linkText)) {
                console.log(`Auction ${auctionDate}, ${linkText} already exists. Skipping.`);
                continue;
            }

            try {
                const downloadedFileName = `saledocs/${linkText}`
                if (!existsSync(downloadedFileName)) {
                    await download_pdf(`https://www.nycourts.gov${pdfUrl}`, downloadedFileName);
                }
                console.log(`Extracting text from: ${downloadedFileName}`);

                const extractedText = await extractTextFromPdf(downloadedFileName);
                let indexNumber = await extractIndexNumber(extractedText)
                if(isInteractive && !indexNumber) {
                    const child = exec(`open "${downloadedFileName}"`);

                    // Get 'block' and 'lot' from the user
                    console.log("\nOpening pdf file")
                    indexNumber = await prompt('Enter index #: ');

                    // Close the PDF
                    exec(`osascript -e 'tell application "Preview" to close (every document whose path is "${pdfPath}")'`);
                }

                let [block, lot] = extractBlockLot(extractedText);
                if (isInteractive && (!block || !lot)) {
                    // Open the PDF file with the default application on macOS
                    const child = exec(`open "${pdfPath}"`);

                    // Get 'block' and 'lot' from the user
                    console.log("\nOpening pdf file")
                    block = await prompt('Enter block: ');
                    lot = await prompt('Enter lot: ');

                    // Close the PDF
                    exec(`osascript -e 'tell application "Preview" to close (every document whose path is "${pdfPath}")'`);
                }
                const lien = extractJudgement(extractedText)
                const address = convertToAddress(linkText);

                stringifier.write(['Brooklyn', auctionDate, indexNumber, address, block, lot, lien])
                console.log(`Added new auction data for ${linkText}`);
                if (indexNumber) {
                    const newFileName = "saledocs/" + indexNumber.replace('/','-') + '.pdf'
                    await copyFile(downloadedFileName, newFileName)
                }
                
                await unlink(downloadedFileName);

            } catch (error) {
                console.error(`Error processing ${linkText}: ${error.message}`);
                // Optionally, you can add a retry mechanism here 
            }
        }
    } finally {
        await browser.close();
        stringifier.end(); // End the stringifier stream properly
    }
}

main().catch(err => console.error(err));